packages/react-aria-components/src/Toolbar.tsx

```diff
@@ -0,0 +1,176 @@
+import {useLocale} from '@react-aria/i18n';
+
+export interface ToolbarRenderProps {
+  /**
+   * Whether the dropzone is currently hovered with a mouse.
+   * @selector [data-orientation]
+   */
+  orientation: Orientation
+}
+
+export interface ToolbarProps extends MultipleSelection, SlotProps, RenderProps<ToolbarRenderProps> {
+  /** Accessible label for the Toolbar. */
+  'aria-label'?: string,
+
+  /** Accessible label for the Toolbar. */
+  'aria-labelledby'?: string,
+
+  /**
+   * The orientation of the entire toolbar.
+   * @default 'horizontal'
+   */
+  orientation?: 'horizontal' | 'vertical'
+}
+
+export const ToolbarContext = createContext<ContextValue<ToolbarProps, HTMLDivElement>>({});
+
+// Extract out to aria hooks?
+// Any reason not to include ref? i feel like we've been sad about not including it down the line
+// same with state...
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function useToolbar(props: ToolbarProps & {isInsideAToolbar: boolean}, ref: ForwardedRef<HTMLDivElement>) {
+  const {
+    'aria-label': ariaLabel,
+    'aria-labelledby': ariaLabelledBy,
+    orientation = 'horizontal',
+    isInsideAToolbar
+  } = props;
+  const {direction} = useLocale();
+  const shouldReverse = direction === 'rtl' && orientation === 'horizontal';
+  const focusManager = useFocusManager();
+
+  const onKeyDown: KeyboardEventHandler = (e) => {
+    if (
+      (orientation === 'horizontal' && e.key === 'ArrowRight')
+      || (orientation === 'vertical' && e.key === 'ArrowDown')) {
+      if (shouldReverse) {
+        focusManager.focusPrevious();
+      } else {
+        focusManager.focusNext();
+      }
+    } else if (
+      (orientation === 'horizontal' && e.key === 'ArrowLeft')
+      || (orientation === 'vertical' && e.key === 'ArrowUp')) {
+      if (shouldReverse) {
+        focusManager.focusNext();
+      } else {
+        focusManager.focusPrevious();
+      }
+    } else if (e.key === 'Tab') {
+      // When the tab key is pressed, we want to move focus
+      // out of the entire toolbar. To do this, move focus
+      // to the first or last focusable child, and let the
+      // browser handle the Tab key as usual from there.
+      e.stopPropagation();
+      lastFocused.current = document.activeElement as HTMLElement;
+      if (e.shiftKey) {
+        focusManager.focusFirst();
+      } else {
+        focusManager.focusLast();
+      }
+      return;
+    } else {
+      // if we didn't handle anything, return early so we don't preventDefault
+      return;
+    }
+
+    // Prevent arrow keys from being handled by nested action groups.
+    e.stopPropagation();
+    e.preventDefault();
+  };
+
+  // Record the last focused child when focus moves out of the toolbar.
+  const lastFocused = useRef<HTMLElement | null>(null);
+  const onBlur = (e) => {
+    if (!e.currentTarget.contains(e.relatedTarget) && !lastFocused.current) {
+      lastFocused.current = e.target;
+    }
+  };
```

Comment:
I suppose this is to handle cases where we have non-ActionGroup components in the Toolbar? Otherwise, the roving tabindex should be sufficient.

Reply:
Ah I see the RAC example now, makes sense